modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java

```diff
@@ -0,0 +1,125 @@
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.internal.replicator.exception.ReplicationError;
+import org.apache.ignite.internal.replicator.message.CompleteOperationRequest;
+import org.apache.ignite.internal.replicator.message.ReplicaMessagesFactory;
+import org.apache.ignite.internal.replicator.message.ReplicaRequest;
+import org.apache.ignite.internal.replicator.message.ReplicaResponse;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.apache.ignite.lang.IgniteStringFormatter;
+
+/**
+ * Replica server.
+ * TODO:IGNITE-17257 Implement Replica server-side logic.
+ */
+public class Replica {
+    /** Replicator network message factory. */
+    private static final ReplicaMessagesFactory REPLICA_MESSAGES_FACTORY = new ReplicaMessagesFactory();
+
+    /** Replica group identity, this id is the same as the considered partition's id. */
+    private final String replicaGrpId;
+
+    /** Replica listener. */
+    private final ReplicaListener listener;
+
+    /**
+     * The map matches an operation id to the future.
+     */
+    private final ConcurrentHashMap<UUID, CompletableFuture> ops = new ConcurrentHashMap<>();
+
+    /**
+     * The constructor of replica server.
+     *
+     * @param replicaGrpId Replication group id.
+     * @param listener Replica listener.
+     */
+    public Replica(
+            String replicaGrpId,
+            ReplicaListener listener
+    ) {
+        this.replicaGrpId = replicaGrpId;
+        this.listener = listener;
+    }
+
+    /**
+     * Process a replication request on the replica.
+     *
+     * @param request Request to replication.
+     * @return Response.
+     */
+    public ReplicaResponse processRequest(ReplicaRequest request) { // define proper set of exceptions that might be thrown.
+        assert replicaGrpId.equals(request.replicationGroupId()) : IgniteStringFormatter.format(
+                "Partition mismatch: request does not match the replica [reqPartId={}, replicaGrpId={}]", request.replicationGroupId(),
+                replicaGrpId);
+
+        //TODO:IGNITE-17378 Check replica is alive.
+
+        ReplicaResponse response;
+
+        try {
+            if (request instanceof CompleteOperationRequest) {
+                var opsRequest = (CompleteOperationRequest) request;
+
+                List<UUID> opIds = opsRequest.operationIds();
+
+                HashMap<UUID, Object> opRes = null;
+
+                if (opIds != null && !opIds.isEmpty()) {
+                    opRes = new HashMap<>(opIds.size());
+
+                    for (Entry<UUID, CompletableFuture> op : ops.entrySet()) {
+                        if (opIds.contains(op.getKey())) {
+                            opRes.put(op.getKey(), op.getValue().join());
+
+                            ops.remove(op.getKey());
```

Comment:
Who and how will remove operation futures in case of coordinator failure?

Reply:
I think, it should be solve in transaction recovery procedure

Reply:
I added another request/response procedure for clearing the futures.
Now we can invoke a method ReplicaService#clearResult for clearing remote replica explicitly.